src/ranged.cpp

```diff
@@ -929,33 +1035,50 @@ static int do_aim( player &p, const std::vector<Creature *> &t, int cur_target,
+
+    double missed_by_tiles = accuracy * occupied_tile_fraction;
+
+    //          T = (2*D**2 * (1 - cos V)) ** 0.5   (from iso_tangent)
+    //      cos V = 1 - T**2 / (2*D**2)
+    double cosV = 1 - missed_by_tiles * missed_by_tiles / (2 * range * range);
+    double shot_dispersion = ( cosV > 1.0 ? M_PI : acos( cosV ) ) * 180 * 60 / M_PI;
+    double sigma = dispersion / dispersion_sigmas;
+
+    // erf(x / (sigma * sqrt(2))) is the probability that a measurement
+    // of a normally distributed variable with standard deviation sigma
+    // lies in the range [-x..x]
+    return std::erf( shot_dispersion / ( sigma * M_SQRT2 ) );
+}
+
 static int print_aim_bars( const player &p, WINDOW *w, int line_number, item *weapon,
                            Creature *target, double predicted_recoil ) {
     // This is absolute accuracy for the player.
     // TODO: push the calculations duplicated from Creature::deal_projectile_attack() and
     // Creature::projectile_attack() into shared methods.
     // Dodge is intentionally not accounted for.
 
-    // Confidence is chance of the actual shot being under the target threshold,
-    // This simplifies the calculation greatly, that's intentional.
-    const double aim_level = p.get_weapon_dispersion( *weapon ) + predicted_recoil + p.recoil_vehicle();
+    const double dispersion = p.get_weapon_dispersion( *weapon ) + predicted_recoil + p.recoil_vehicle();
     const double range = rl_dist( p.pos(), target->pos() );
-    const double missed_by = aim_level * 0.00021666666666666666 * range;
-    const double hit_rating = missed_by;
-    const double confidence = 1 / hit_rating;
+
+    const double p_headshot = p.projectile_attack_chance( dispersion, range, accuracy_headshot );
+    const double p_goodhit = p.projectile_attack_chance( dispersion, range, accuracy_goodhit );
+    const double p_grazing = p.projectile_attack_chance( dispersion, range, accuracy_grazing );
+
     // This is a relative measure of how steady the player's aim is,
     // 0 it is the best the player can do.
     const double steady_score = predicted_recoil - p.effective_dispersion( p.weapon.sight_dispersion() );
     // Fairly arbitrary cap on steadiness...
     const double steadiness = 1.0 - steady_score / MIN_RECOIL;
 
     const std::array<std::pair<double, char>, 3> confidence_ratings = {{
-        std::make_pair( accuracy_headshot, '*' ),
-        std::make_pair( accuracy_goodhit, '+' ),
-        std::make_pair( accuracy_grazing, '|' ) }};
+        std::make_pair( p_headshot, '*' ),
+        std::make_pair( p_goodhit, '+' ),
+        std::make_pair( p_grazing, '|' ) }};
 
     const int window_width = getmaxx( w ) - 2; // Window width minus borders.
-    const std::string &confidence_bar = get_labeled_bar( confidence, window_width, _( "Confidence" ),
+    const std::string &confidence_bar = get_labeled_bar( 1.0, window_width, _( "Confidence" ),
```

Comment:
get_labeled_bar puts the transition points at (val * bar_width * value from the iterator).
But the values in the iterators are already the fraction of the bar width I want to use, so there's no use to that extra scale factor.